libraries/lib-effects/EffectOutputTracks.cpp

```diff
@@ -35,24 +36,32 @@ EffectOutputTracks::EffectOutputTracks(
          : dynamic_cast<const WaveTrack*>(pTrack) && pTrack->GetSelected();
       };
 
-   using namespace BasicUI;
-   auto progress = MakeProgress(
-      XO("Pre-processing"), XO("Rendering Time-Stretched Audio"),
-      ProgressShowCancel);
-   const auto numTracks = trackRange.size();
-   auto count = 0;
-   auto reportProgress = [&](double progressFraction) {
-      const auto overallProgress = (count + progressFraction) / numTracks;
-      const auto result = progress->Poll(overallProgress * 1000, 1000);
-      if (result != ProgressResult::Success)
-         throw UserException{};
-   };
    for (auto aTrack : trackRange) {
-      auto list = aTrack->Duplicate(effectTimeInterval, reportProgress);
+      auto list = aTrack->Duplicate();
+      if (effectTimeInterval.has_value() && effectTimeInterval->second > effectTimeInterval->first)
+      {
+         using namespace BasicUI;
+         auto progress = MakeProgress(
+            XO("Pre-processing"), XO("Rendering Time-Stretched Audio"),
+            ProgressShowCancel);
+         const auto waveTracks = list->Any<WaveTrack>();
+         const auto numTracks = waveTracks.size();
+         auto count = 0;
+         auto reportProgress = [&](double progressFraction) {
+            const auto overallProgress = (count + progressFraction) / numTracks;
+            const auto result = progress->Poll(overallProgress * 1000, 1000);
+            if (result != ProgressResult::Success)
+               throw UserException {};
+         };
+         for (const auto& track : waveTracks)
+         {
+            track->ApplyStretchRatio(effectTimeInterval, reportProgress);
+            ++count;
+         }
+      }
       mIMap.push_back(aTrack);
       mOMap.push_back(*list->begin());
       mOutputTracks->Append(std::move(*list));
-      ++count;
```

Comment:
Instead you could move stretching after this loop.  Count wave tracks or maybe only the tracks needing a stretch first and construct a progress indicator with that denominator.  Then another loop to do stretching.

Reply:
Sounds better. Trying that ...